### DISCUSSION TITLE
PDN generation

### DISCUSSION THREAD

**Discussion by armleo** [The-OpenROAD-Project/OpenLane#950] [General]
I have pins that deliver power to Macro. How to force OpenLane to connect the pin to the internal ring? I can't just reduce the size of the macro and then put the rings outside the macro, since that layers are busy. The digital pin locations are fixed, since the macros are fixed. Same for power pins (see the light blue big object on the right). How do I force OpenLane to connect the power to the pin?

![image](https://user-images.githubusercontent.com/7474152/155172737-21304722-a094-44fc-851a-028bbbde9f55.png)

**Comment by maliberty** [comment]
You need to expose the power in the LEF macro for the block and then use a custom PDN_CFG based on https://openlane-docs.readthedocs.io/en/latest/hardening_macros.html?highlight=PDN_CFG#id13

**Comment by armleo** [comment]
I cant run openlane on analog top level. And the guide is for integrating existing macro into openlane design. What I am trying to do is connect the pin of vccd1 (light blue bug blob on the right) to the straps. But it does not use my def's power straps, instead it just removes them in apply def template and creates custom one.

Do I need to just set the margins to zero and then connect the rings? I think I have to do it that way.
